test/new_relic/agent/log_event_aggregator_test.rb

```diff
@@ -521,5 +527,164 @@ def test_records_log_events_not_within_default_severities
         assert_equal(log_message, events.first.last['message'])
       end
     end
+
+    def test_record_json_sets_severity_when_given_level
+      @aggregator.record_logstasher_event({'level' => :warn, 'message' => 'yikes!'})
+      _, events = @aggregator.harvest!
+
+      assert_equal :warn, events[0][1]['level']
+      assert_metrics_recorded([
+        'Logging/lines/warn'
+      ])
+    end
+
+    def test_record_json_sets_severity_unknown_when_no_level
+      @aggregator.record_logstasher_event({'message' => 'hello there'})
+      _, events = @aggregator.harvest!
+
+      assert_equal 'UNKNOWN', events[0][1]['level']
+      assert_metrics_recorded([
+        'Logging/lines/UNKNOWN'
+      ])
+    end
+
+    def test_record_json_does_not_record_if_message_is_nil
+      @aggregator.record_logstasher_event({'level' => :info, 'message' => nil})
+      _, events = @aggregator.harvest!
+
+      assert_empty events
+    end
+
+    def test_record_json_does_not_record_if_message_empty_string
+      @aggregator.record_logstasher_event({'level' => :info, 'message' => ''})
+      _, events = @aggregator.harvest!
+
+      assert_empty events
+    end
+
+    def test_record_json_returns_message_when_avaliable
+      @aggregator.record_logstasher_event({'level' => :warn, 'message' => 'A trex is near'})
+      _, events = @aggregator.harvest!
+
+      assert_equal 'A trex is near', events[0][1]['message']
+    end
+
+    def test_record_json_returns_empty_string_when_unavaliable
+      @aggregator.record_logstasher_event({'level' => :info})
+      _, events = @aggregator.harvest!
+
+      assert_equal '', events[0][1]['message']
+    end
```

Comment:
I think this is [a departure from the spec.](https://source.datanerd.us/agents/agent-specs/blob/main/Application-Logging.md#nodejs-exceptions) I believe this attribute is supposed to be omitted if it is nil, empty, missing or simply whitespace.

Why does a missing "message" key create one equal to an empty string, whereas a "message" key equal to an empty string is not sent at all?

The same behavior is [tested here](https://github.com/newrelic/newrelic-ruby-agent/pull/2559/files#diff-b95067722289f3e6cbeef3d275c924f8900cbf9ea650450eecc910ac951e75deR61).

Reply:
Oohh you're absolutely right. On re-reading the spec, we shouldn't be recording a log without a message unless it supports a mapped context functionality. Logstasher logs don't have a message (the only time they do is when a user adds their own log statement). But they do have a lot of other contextual information about events, so I think we could consider it falling into the supporting context attributes. Although this may be worth talking about further live.

For now, I've stopped recording a message when one isn't avaliable. https://github.com/newrelic/newrelic-ruby-agent/pull/2559/commits/f689fb7b0444d05e8a7f1acd284a48e710994fc9

Reply:
That sounds good to me! Happy to talk more about it live if you'd like.